src/data/index.ts

```diff
@@ -23,11 +23,125 @@ const BACKEND_GET = "https://json.excalidraw.com/api/v1/";
+
 // TODO: Defined globally, since file handles aren't yet serializable.
 // Once `FileSystemFileHandle` can be serialized, make this
 // part of `AppState`.
 (window as any).handle = null;
 
+function byteToHex(byte: number): string {
+  return `0${byte.toString(16)}`.slice(-2);
+}
+
+async function generateRandomID() {
+  var arr = new Uint8Array(20);
+  window.crypto.getRandomValues(arr);
+  return Array.from(arr, byteToHex).join("");
+}
+
+async function generateEncryptionKey() {
+  const key = await window.crypto.subtle.generateKey(
+    {
+      name: "AES-GCM",
+      length: 128,
+    },
+    true, // extractable
+    ["encrypt", "decrypt"],
+  );
+  return (await window.crypto.subtle.exportKey("jwk", key)).k;
+}
+
+export function getCollaborationLinkData(link: string) {
+  if (link.length === 0) {
+    return;
+  }
+  const hash = new URL(link).hash;
+  return hash.match(/^#room_id=([a-zA-Z0-9_-]+),([a-zA-Z0-9_-]+)$/);
+}
+
+export async function generateCollaborationLink() {
+  const id = await generateRandomID();
+  const key = await generateEncryptionKey();
+  return `${window.location.href}#room_id=${id},${key}`;
+}
+
+async function getImportedKey(key: string, usage: string): Promise<CryptoKey> {
+  return await window.crypto.subtle.importKey(
+    "jwk",
+    {
+      alg: "A128GCM",
+      ext: true,
+      k: key,
+      key_ops: ["encrypt", "decrypt"],
+      kty: "oct",
+    },
+    {
+      name: "AES-GCM",
+      length: 128,
+    },
+    false, // extractable
+    [usage],
+  );
+}
+
+export async function decryptSocketUpdateData(
+  encryptedData: ArrayBuffer,
+  key: string,
+) {
+  let elements: readonly ExcalidrawElement[] = [];
+  let appState: AppState = getDefaultAppState();
+  try {
+    const importedKey = await getImportedKey(key, "decrypt");
+
+    const iv = new Uint8Array(12);
+    const decrypted = await window.crypto.subtle.decrypt(
+      {
+        name: "AES-GCM",
+        iv: iv,
+      },
+      importedKey,
+      encryptedData,
+    );
+    // We need to convert the decrypted array buffer to a string
+    const decodedData = new window.TextDecoder("utf-8").decode(
+      new Uint8Array(decrypted) as any,
+    );
+    const data = JSON.parse(decodedData);
+    elements = data.elements || elements;
+    appState = data.appState || appState;
+  } catch (error) {
+    window.alert(t("alerts.decryptSyncFailed"));
+    console.error(error);
+  } finally {
+    return restore(elements, appState, { scrollToContent: true });
+  }
+}
+
+export async function encryptSocketUpdateData(
+  elements: readonly ExcalidrawElement[],
+  appState: AppState,
+  key: string,
+): Promise<ArrayBuffer> {
+  const json = serializeAsJSON(elements, appState);
+  const encoded = new TextEncoder().encode(json);
+  const importedKey = await getImportedKey(key, "encrypt");
+
+  // The iv is set to 0. We are never going to reuse the same key so we don't
+  // need to have an iv. (I hope that's correct...)
```

Comment:
You --are-- reusing the same key to encrypt all the messages... I believe that this basically breaks the encryption scheme. We need to do a review of the protocol with an expert.

Reply:
Ok, so what you need to do in our case is to generate a random 12 digit number, use it to encrypt, pass it in clear to the server and pass it back to the other client, and use it to decrypt it.

This ensures that if you encrypt many messages with the same key, the attacker cannot deduce things from it.